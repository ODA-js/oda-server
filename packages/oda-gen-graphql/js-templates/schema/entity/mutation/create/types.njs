<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-create-types'#>
<#@ context 'entity'#>

<#- chunkStart(`./mutations/create/create${entity.name}Input.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input create#{entity.name}Input {
<#- for (let field of entity.create){#>
      #{field.name}: #{field.type}
<#-}#>
<#- for (let rel of entity.relations.filter(f=>f.persistent)){
const refName = rel.fields.length > 0 ? `embed${rel.ref.entity}CreateInto${entity.name}${rel.cField}Input` : `embed${rel.ref.entity}Input`;
#>
      #{rel.field}: <#if(!rel.single){#>[<#}#>#{refName}<#if(!rel.single){#>]<#}#>
<#-}#>
    }
  `,
});

<#- chunkStart(`./mutations/create/create${entity.name}Payload.ts`); -#>

import { Type } from '../../../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type create#{entity.name}Payload {
      #{entity.payloadName}: #{entity.plural}Edge
    }
  `,
});
