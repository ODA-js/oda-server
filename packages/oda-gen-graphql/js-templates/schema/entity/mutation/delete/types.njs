<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-delete-types'#>
<#@ context 'entity'#>

<#- chunkStart(`./mutations/delete/delete${entity.name}Input.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input delete#{entity.name}Input {
<#- for (let field of entity.unique){#>
      #{field.name}: #{field.type}
<#-}#>
    } 
  `,
});

<#- chunkStart(`./mutations/delete/delete${entity.name}Payload.ts`); -#>

import { Type } from '../../../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type delete#{entity.name}Payload {
      deletedItemId: ID
      #{entity.payloadName}: #{entity.name}
    }
  `,
});
