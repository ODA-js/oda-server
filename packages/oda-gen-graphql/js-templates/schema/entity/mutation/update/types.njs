<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-update-types'#>
<#@ context 'entity'#>

<#- chunkStart(`./mutations/update/update${entity.name}Input.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input update#{entity.name}Input {
    <#- for (let field of entity.update){#>
      #{field.name}: #{field.type}
    <#-}#>
    <#- for (let rel of entity.relations.filter(f=>f.persistent)){

    const createName = rel.fields.length > 0 ? `embed${rel.ref.entity}CreateInto${entity.name}${rel.cField}Input` : `create${rel.ref.entity}Input`;
    const updateName = rel.fields.length > 0 ? `embed${rel.ref.entity}UpdateInto${entity.name}${rel.cField}Input` : `embed${rel.ref.entity}Input`;
    #>
      #{rel.field}: <#if(!rel.single){#>[<#}#>#{updateName}<#if(!rel.single){#>]<#}#>
<#- if(!rel.embedded){#>
      #{rel.field}Unlink: <#if(!rel.single){#>[<#}#>#{updateName}<#if(!rel.single){#>]<#}#>
      #{rel.field}Create: <#if(!rel.single){#>[<#}#>#{createName}<#if(!rel.single){#>]<#}#>
<#-}#>
    <#-}#>
    }
  `,
});

<#- chunkEnd(); -#>

<#- for (let rel of entity.relations.filter(f=>f.persistent && f.fields.length > 0)){
const createName = `embed${rel.ref.entity}CreateInto${entity.name}${rel.cField}Input`;
const updateName = `embed${rel.ref.entity}UpdateInto${entity.name}${rel.cField}Input`;
let eEntity = rel.ref.eEntity;
  let eRels = eEntity.relations;
#>

<#- chunkStart(`./mutations/create/${createName}.ts`); -#>
<# slot('import-embed-entity-create-rel-mutation-types', createName) #>
<# slot('use-embed-entity-create-rel-mutation-types', createName) #>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
input #{createName} {
<#- for (let field of eEntity.create){#>
  #{field.name}: #{field.type}
<#-}#>
<#- for (let field of rel.fields){#>
  #{field.name}: #{field.type}
<#-}#>
<#-
 for (let eRel of eRels.filter(f=>f.persistent)){
const refName = eRel.fields.length > 0 ? `embed${eRel.ref.entity}UpdateInto${eEntity}${eRel.cField}Input` : `embed${eRel.ref.entity}Input`;
#>
  #{eRel.field}: <#if(!eRel.single){#>[<#}#>#{refName}<#if(!eRel.single){#>]<#}#>
<#-}#>
}`
});

<#- chunkEnd(); -#>
<#- chunkStart(`./mutations/update/${updateName}.ts`); -#>
<# slot('import-embed-entity-update-rel-mutation-types', updateName) #>
<# slot('use-embed-entity-update-rel-mutation-types', updateName) #>

import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
input #{updateName} {
<#- for (let field of eEntity.update){#>
  #{field.name}: #{field.type}
<#-}#>
<#- for (let field of rel.fields){#>
  #{field.name}: #{field.type}
<#-}#>
<#-
 for (let eRel of eRels.filter(f=>f.persistent)){
const refName = eRel.fields.length > 0 ? `embed${eRel.ref.entity}UpdateInto${eEntity}${eRel.cField}Input` : `embed${eRel.ref.entity}Input`;
#>
  #{eRel.field}: <#if(!eRel.single){#>[<#}#>#{refName}<#if(!eRel.single){#>]<#}#>
<#-}#>
}
`});
<#- chunkEnd(); -#>

<#-}#>


<#- chunkStart(`./mutations/update/update${entity.name}Payload.ts`); -#>

import { Type } from '../../../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type update#{entity.name}Payload {
      #{entity.payloadName}: #{entity.name}
    }
  `,
});
