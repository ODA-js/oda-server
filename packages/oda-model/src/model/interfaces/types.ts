export type RelationType = 'HasMany' | 'HasOne' | 'BelongsToMany' | 'BelongsTo';

export type MetaModelType =
  | 'metadata'
  | 'model'
  | 'query'
  | 'mutation'
  | 'package'
  | 'entity'
  | 'scalar'
  | 'mixin'
  | 'union'
  | 'enum'
  | 'enum-item'
  | 'field-base'
  | 'field'
  | 'enum-field'
  | 'entity-field'
  | 'relation-field'
  | 'relation'
  | 'operation'
  | 'ref'
  | RelationType;

export type Multiplicity = 'one' | 'many';
/**
 * kind of complex type
 */
export type ComplexTypeKind = 'enum' | 'entity';

export type OperationKind = 'create' | 'read' | 'update' | 'delete';

export type DirectiveLocation =
  | 'QUERY'
  | 'MUTATION'
  | 'SUBSCRIPTION'
  | 'FIELD'
  | 'FRAGMENT_DEFINITION'
  | 'FRAGMENT_SPREAD'
  | 'INLINE_FRAGMENT'
  | 'SCHEMA'
  | 'SCALAR'
  | 'OBJECT'
  | 'FIELD_DEFINITION'
  | 'ARGUMENT_DEFINITION'
  | 'INTERFACE'
  | 'UNION'
  | 'ENUM'
  | 'ENUM_VALUE'
  | 'INPUT_OBJECT'
  | 'INPUT_FIELD_DEFINITION';

export interface INamed {
  readonly name: string;
}

export interface FieldArgsForHash {
  type?: string;
  required?: boolean;
  defaultValue?: string;
}

export interface FieldArgs extends INamed, FieldArgsForHash {}
/**
 * Complext type can refer to existing type system
 */
export type EntityType = {
  /**
   * complext type name
   */
  name: string;
  /**
   * kind of complex type
   */
  type: 'entity';
  /**
   * multiplicity of it
   */
  multiplicity: Multiplicity;
};

export type EnumType = {
  /**
   * complext type name
   */
  name: string;
  /**
   * kind of complex type
   */
  type: 'enum';
  /**
   * multiplicity of it
   */
  multiplicity: Multiplicity;
};

export type FieldType = string | EnumType | EntityType;

export interface AsHash<T extends INamed> {
  [name: string]: T;
}

export function HashToMap<T extends INamed>(input: AsHash<T>): Map<string, T> {
  const res = Object.keys(input).map(name => {
    return [name, input[name] as T] as [string, T];
  });
  return new Map(res);
}

export function MapToHash<T extends INamed>(input: Map<string, T>): AsHash<T> {
  return [...input.entries()].reduce(
    (hash, [name, value]) => {
      hash[name] = value;
      return hash;
    },
    {} as AsHash<T>,
  );
}
