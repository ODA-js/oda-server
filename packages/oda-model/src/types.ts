import { readFileSync, writeFileSync } from 'fs';
import { IUpdatableBase } from './element';
import { get, set, has } from 'lodash';

export type Nullable<T> = { [P in keyof T]: T[P] | undefined | null };

/**
 *
 */
export type assignInput<S, I, V = any> = {
  /**
   * source object
   */
  src: S;
  /**
   * dest object
   */
  input: Nullable<I>;
  /**
   * field from source object which is assigned
   */
  field?: keyof S | keyof S & keyof I | string;
  /**
   * function that will call to assign value
   */
  effect?: (src: S, value: V) => void;
  /**
   * function that check if effect is allowed
   */
  allowEffect?: (src: S, value: V) => boolean;
  /**
   * is field required, so it can't be reset by set null.
   * in this case assign will throw
   */
  required?: boolean;
  /**
   * field from dest object which value will be assigned to dest field
   */
  inputField?: keyof I;
  /**
   * function that will call to set default value
   */
  setDefault?: (src: S) => void;
};

/**
 * check specified value from input and assign
 * @param options assign Options
 */
export function assignValue<S extends object, I extends object, V = any>(
  options: assignInput<S, I, V>,
) {
  let {
    src,
    input,
    field,
    effect,
    required,
    inputField,
    setDefault,
    allowEffect,
  } = options;

  if (!inputField) {
    inputField = field as keyof I;
  }
  if (has(input, inputField)) {
    // update value
    if (get(input, inputField) != null) {
      if (effect) {
        if (
          (allowEffect && allowEffect(src, get(input, inputField) as any)) ||
          !allowEffect
        ) {
          effect(src, get(input, inputField) as any);
        } else if (
          allowEffect &&
          !allowEffect(src, get(input, inputField) as any) &&
          setDefault instanceof Function
        ) {
          setDefault(src);
        }
      } else if (field) {
        set(src, field, get(input, inputField));
      }
    } else if (inputField && get(input, inputField) === null && required) {
      if (setDefault instanceof Function) {
        setDefault(src);
      } else if (field) {
        throw new Error("can't reset required value");
      }
    } else if (inputField && get(input, inputField) === null) {
      if (setDefault instanceof Function) {
        setDefault(src);
      } else if (field) {
        set(src, field, undefined);
      }
    }
  } else {
    // set default value
    if (field && get(src, field) == null) {
      if (setDefault instanceof Function) {
        setDefault(src);
      }
    }
  }
}

export type RelationType = 'HasMany' | 'HasOne' | 'BelongsToMany' | 'BelongsTo';

export type MetaModelType =
  | 'element'
  | 'metadata'
  | 'package-base'
  | 'model'
  | 'package'
  | 'entity'
  | 'mixin'
  | 'query'
  | 'mutation'
  | 'scalar'
  | 'union'
  | 'directive'
  | 'enum'
  | 'enum-item'
  | 'entity-base'
  | 'field-base'
  | 'simple-field'
  | 'enum-field'
  | 'entity-field'
  | 'relation-field'
  | 'relation'
  | 'relation-base'
  | 'operation'
  | 'ref'
  | 'args'
  | 'model-hook'
  | RelationType;

export type Multiplicity = 'one' | 'many';
/**
 * kind of complex type
 */
export type ComplexTypeKind = 'enum' | 'entity';

export type OperationKind =
  | 'create'
  | 'readOne'
  | 'readMany'
  | 'update'
  | 'delete'
  | 'addTo'
  | 'removeFrom';

export type DirectiveLocation =
  | 'QUERY'
  | 'MUTATION'
  | 'SUBSCRIPTION'
  | 'FIELD'
  | 'FRAGMENT_DEFINITION'
  | 'FRAGMENT_SPREAD'
  | 'INLINE_FRAGMENT'
  | 'SCHEMA'
  | 'SCALAR'
  | 'OBJECT'
  | 'FIELD_DEFINITION'
  | 'ARGUMENT_DEFINITION'
  | 'INTERFACE'
  | 'UNION'
  | 'ENUM'
  | 'ENUM_VALUE'
  | 'INPUT_OBJECT'
  | 'INPUT_FIELD_DEFINITION';

export interface INamed {
  readonly name: string;
}

/**
 * Complext type can refer to existing type system
 */
export type EntityType = {
  /**
   * complext type name
   */
  name: string;
  /**
   * kind of complex type
   */
  type: 'entity';
  /**
   * multiplicity of it
   */
  multiplicity?: Multiplicity;
};

export type EnumType = {
  /**
   * complex type name
   */
  name: string;
  /**
   * kind of complex type
   */
  type: 'enum';
  /**
   * multiplicity of it
   */
  multiplicity?: Multiplicity;
};

export type FieldType = string | EnumType | EntityType;

export interface AsHash<T extends Partial<INamed> & object> {
  [name: string]: T;
}

export type NamedArray<T extends INamed> = Array<T>;

function mergeDupes<O extends Partial<INamed> & object>(
  src: Map<string, O>,
  items: [string, O][],
  killDupes?: (obj: O, src: O) => void,
) {
  items.forEach(item => {
    if (killDupes) {
      const dupe = src.get(item[0]);
      if (dupe) {
        killDupes(dupe, item[1]);
      } else {
        src.set(item[0], item[1]);
      }
    } else {
      src.set(item[0], item[1]);
    }
  });
}

export function ArrayToMap<T extends INamed, O extends INamed = T>(
  input: NamedArray<T>,
  process?: (v: T, index: number) => O,
  killDupes?: (obj: O, src: O) => void,
) {
  const res = new Map<string, O>();
  mergeDupes(
    res,
    input.map((f, index) => {
      let result: O = (f as unknown) as O;
      if (process) {
        result = (process(f, index) as unknown) as O;
      }
      return [result.name, result] as [string, O];
    }),
    killDupes,
  );
  return res;
}

export function HashToMap<
  T extends Partial<INamed> & object,
  O extends Partial<INamed> & object = T
>(
  input: AsHash<T>,
  process?: (name: string, v: T) => O,
  killDupes?: (obj: O, src: O) => void,
): Map<string, O> {
  const res = new Map<string, O>();
  mergeDupes(
    res,
    Object.keys(input).map(key => {
      let result: O | T = (input[key] as unknown) as O;
      if (process) {
        result = process(key, (result as unknown) as T);
      } else {
        result = {
          name: key,
          ...result,
        };
      }
      return [result.name, result] as [string, O];
    }),
    killDupes,
  );
  return res;
}

export function HashToArray<T extends Partial<INamed> & object>(
  input: AsHash<T>,
  process?: (name: string, v: T) => T,
  killDupes?: (obj: T, src: T) => void,
): T[] {
  const res = new Map<string, T>();
  mergeDupes(
    res,
    Object.keys(input).map(key => {
      let result = input[key];
      if (process) {
        result = process(key, result);
      } else {
        result = {
          ...result,
          name: key,
        };
      }
      return [result.name, result] as [string, T];
    }),
    killDupes,
  );
  return [...res.values()];
}

export function ArrayToHash<T extends INamed>(
  input: Array<T>,
  process?: (v: T) => T,
  killDupes?: (obj: T, src: T) => void,
): AsHash<T> {
  const res = new Map<string, T>();
  mergeDupes(
    res,
    input.map(f => {
      let result: T = (f as unknown) as T;
      if (process) {
        result = (process(f) as unknown) as T;
      }
      return [result.name, result] as [string, T];
    }),
    killDupes,
  );
  return [...res.values()].reduce(
    (result, item) => {
      result[item.name] = item;
      return result;
    },
    {} as AsHash<T>,
  );
}

export function MapToHash<T extends INamed, O extends INamed = T>(
  input: Map<string, T>,
  process?: (name: string, v: T) => O,
): AsHash<O> {
  return [...input.entries()].reduce(
    (hash, [name, value]) => {
      hash[name] = process
        ? (process(name, value) as O)
        : ((value as unknown) as O);
      return hash;
    },
    {} as AsHash<O>,
  );
}

export function MapToArray<T extends INamed, O extends INamed = T>(
  input: Map<string, T>,
  process?: (name: string, v: T) => O,
): NamedArray<O> {
  return [...input.entries()].reduce(
    (r, curr) => {
      r.push(
        process
          ? (process(curr[0], curr[1]) as O)
          : ((curr[1] as unknown) as O),
      );
      return r;
    },
    [] as O[],
  );
}

export type Merger<I extends INamed> = (
  objValue: any,
  srcValue: any,
  key: keyof I,
  object: I,
  source: I,
  stack: string[],
) => any;
/**
 * it checks and create item
 * @param srcMap external source map
 * @param create constructor for creating new item
 */
export function AssignAndKillDupes<
  T extends IUpdatableBase & INamed,
  I extends INamed
>(srcMap: Map<string, T> | undefined, create: new (input: I) => T) {
  return (i: I | string) => {
    let res: T | undefined;
    /** if item is string */
    if (typeof i === 'string') {
      /** check if we have one in external map */
      res = srcMap && srcMap.get(i);
      if (!res) {
        /** if not-> create it an push to external map */
        res = new create({ name: i } as any);
        srcMap && srcMap.set(res.name, res);
      }
    } else {
      /** if item is input object */
      res = new create(i);
      /** if we have the same in external map */
      const original = srcMap && srcMap.get(res.name);
      if (original) {
        /** if we have have then merge it with this one*/
        original.mergeWith(res.toObject());
      } else {
        /** set as original one */
        srcMap && srcMap.set(res.name, res);
      }
    }
    return res;
  };
}

export function loadFromFile<T, I>(
  filename: string,
  create: new (inp: I) => T,
) {
  const txt = readFileSync(filename);
  return new create(JSON.parse(txt.toString()) as I);
}

export function saveToFile<T extends { toObject: () => any }>(
  filename: string,
  inp: T,
) {
  writeFileSync(filename, inp.toObject());
}
