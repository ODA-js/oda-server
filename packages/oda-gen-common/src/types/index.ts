import { StateOfFieldType } from './stateOfField';
import { StateOfConectionType } from './stateOfConnection';
import { PageInfoType } from './pageInfo';
import { DateType } from './date';
import { IdType } from './id';
import { ImageSizeType } from './imageSize';
import { JSONType } from './json';
import { GQLModule } from './empty';
import { MutationKindType } from './mutationType';
import { WhereBoolean } from './whereBoolean';
import { WhereDate } from './whereDate';
import { WhereFloat } from './whereFloat';
import { WhereID } from './whereID';
import { WhereInt } from './whereInt';
import { WhereString } from './whereString';
import { WhereListOfStrings } from './whereListOfStrings';
import { WhereMutationKind } from './whereMutationKind';
import { WhereJSON } from './whereJSON';

export {
  GQLModule,
  StateOfFieldType,
  StateOfConectionType,
  PageInfoType,
  DateType,
  IdType,
  ImageSizeType,
  JSONType,
  WhereBoolean,
  WhereDate,
  WhereFloat,
  WhereID,
  WhereInt,
  WhereString,
  WhereListOfStrings,
  WhereMutationKind,
  WhereJSON,
};

export class DefaultTypes extends GQLModule {
  protected _name = 'DefaultTypes';
  protected _composite = [
    new StateOfFieldType({}),
    new StateOfConectionType({}),
    new PageInfoType({}),
    new DateType({}),
    new IdType({}),
    new ImageSizeType({}),
    new JSONType({}),
    new MutationKindType({}),
    new WhereBoolean({}),
    new WhereDate({}),
    new WhereFloat({}),
    new WhereID({}),
    new WhereInt({}),
    new WhereString({}),
    new WhereListOfStrings({}),
    new WhereMutationKind({}),
    new WhereJSON({}),
  ];
}
