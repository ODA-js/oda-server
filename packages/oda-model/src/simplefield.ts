import {
  FieldBaseInput,
  FieldBaseInternal,
  FieldBaseMetaInfo,
  IFieldBase,
  FieldBase,
  FieldBasePersistence,
  FieldBaseOutput,
} from './fieldbase';
import { merge, get } from 'lodash';
import { EnumType, Nullable, assignValue, MetaModelType } from './types';
import { Internal } from './element';
``;
export interface ISimpleField
  extends IFieldBase<
    SimpleFieldMeta,
    SimpleFieldInput,
    SimpleFieldPersistence,
    SimpleFieldOutput
  > {
  readonly type: string | EnumType;
  readonly list?: boolean;
  readonly defaultValue?: string;
}

export interface SimpleFieldPersistence extends FieldBasePersistence {
  derived: boolean;
  persistent: boolean;
}

export interface SimpleFieldMeta
  extends FieldBaseMetaInfo<SimpleFieldPersistence> {
  defaultValue?: string;
}

export interface SimpleFieldInternal extends FieldBaseInternal {
  type: string | EnumType;
  list: boolean;
}

export interface SimpleFieldInput
  extends FieldBaseInput<SimpleFieldMeta, SimpleFieldPersistence> {
  type?: string | EnumType;
  defaultValue?: string;
  list?: boolean;
}

export interface SimpleFieldOutput
  extends FieldBaseOutput<SimpleFieldMeta, SimpleFieldPersistence> {
  type?: string | EnumType;
  defaultValue?: string;
  list?: boolean;
}

export const simpleFieldDefaultMetaInfo = {};
export const simpleFieldDefaultInput = { metadata: simpleFieldDefaultMetaInfo };

export class SimpleField
  extends FieldBase<
    SimpleFieldMeta,
    SimpleFieldInput,
    SimpleFieldInternal,
    SimpleFieldPersistence,
    SimpleFieldOutput
  >
  implements ISimpleField {
  public get modelType(): MetaModelType {
    return typeof this[Internal].type === 'string'
      ? 'simple-field'
      : 'enum-field';
  }

  get type(): string | EnumType {
    return this[Internal].type;
  }

  // if it is the field is List of Items i.e. String[]
  get list(): boolean {
    return get(this[Internal], 'list', false);
  }

  get defaultValue(): string | undefined {
    return get(this.metadata, 'defaultValue');
  }

  constructor(init: SimpleFieldInput) {
    super(merge({}, simpleFieldDefaultInput, init));
  }

  public updateWith(input: Nullable<SimpleFieldInput>) {
    super.updateWith(input);

    assignValue<
      SimpleFieldInternal,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['type']>
    >({
      src: this[Internal],
      input,
      field: 'type',
      effect: (src, value) => {
        if (typeof value !== 'string' && !value.multiplicity) {
          value.multiplicity = 'one';
        }
        src.type = value;
      },
      setDefault: src => (src.type = 'string'),
    });

    assignValue<
      SimpleFieldInternal,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['list']>
    >({
      src: this[Internal],
      input,
      field: 'list',
      required: true,
      effect: (src, value) => {
        if (this.modelType === 'enum-field') {
          (src.type as EnumType).multiplicity = value ? 'many' : 'one';
        }
        src.list = value;
      },
      setDefault: src => {
        if (this.modelType === 'enum-field') {
          src.list =
            (src.type as EnumType).multiplicity === 'many' ? true : false;
        } else {
          src.list = false;
        }
      },
    });

    assignValue<
      SimpleFieldMeta,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['derived']>
    >({
      src: this.metadata,
      input,
      inputField: 'derived',
      effect: (src, value) => {
        src.persistence.derived = value;
        if (!value) {
          this.metadata.persistence.persistent = true;
        }
      },
      setDefault: src => (src.persistence.derived = false),
    });

    assignValue<
      SimpleFieldMeta,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['persistent']>
    >({
      src: this.metadata,
      input,
      inputField: 'persistent',
      effect: (src, value) => (src.persistence.persistent = value),
      setDefault: src => (src.persistence.persistent = true),
    });

    assignValue<
      SimpleFieldMeta,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['args']>
    >({
      src: this.metadata,
      input,
      inputField: 'args',
      allowEffect: (_, value) =>
        Array.isArray(value) ? value.length > 0 : Object.keys(value).length > 0,
      effect: (src, _) => (src.persistence.derived = true),
    });

    assignValue<
      SimpleFieldMeta,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['defaultValue']>
    >({
      src: this.metadata,
      input,
      inputField: 'defaultValue',
      effect: (src, value) => {
        src.defaultValue = value;
      },
      allowEffect: (src, _) => !src.persistence.derived,
    });
  }

  public toObject(): SimpleFieldOutput {
    return merge({}, super.toObject(), {
      type: this.type,
      inheritedFrom: this[Internal].inheritedFrom,
      list: this[Internal].list,
    } as Partial<SimpleFieldOutput>);
  }

  public mergeWith(payload: Nullable<SimpleFieldInput>) {
    super.mergeWith(payload);
  }
}
