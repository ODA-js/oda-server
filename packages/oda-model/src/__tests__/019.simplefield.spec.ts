import 'jest';
import { SimpleField } from '../simplefield';
import { EnumType } from '../types';

describe('SimpleField', () => {
  it('default', () => {
    const res = new SimpleField({
      name: 'A',
    });
    expect(res.name).toBe('a');
    expect(res.type).toBe('string');
    expect(res.modelType).toBe('simple-field');
    expect(res.list).toBeFalsy();
    expect(res).toMatchSnapshot('default');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });

  it('default value only for persistent fields', () => {
    const res = new SimpleField({
      name: 'A',
      defaultValue: 'NAME_OF_DEFAULT_VALUE',
    });
    expect(res.name).toBe('a');
    expect(res.defaultValue).toBe('NAME_OF_DEFAULT_VALUE');
  });

  it('default value not for persistent fields', () => {
    const res = new SimpleField({
      name: 'A',
      derived: true,
      defaultValue: 'NAME_OF_DEFAULT_VALUE',
    });
    expect(res.name).toBe('a');
    expect(res.defaultValue).toBeUndefined();
  });

  it('allow persistent derived', () => {
    const res = new SimpleField({
      name: 'A',
      derived: true,
      persistent: true,
      defaultValue: 'NAME_OF_DEFAULT_VALUE',
    });
    expect(res.name).toBe('a');
    expect(res.defaultValue).toBeUndefined();
    expect(res.persistent).toBeTruthy();
    expect(res.derived).toBeTruthy();
  });

  it('default enum', () => {
    const res = new SimpleField({
      name: 'A',
      type: {
        type: 'enum',
        multiplicity: 'many',
        name: 'ENUM1',
      },
    });
    expect(res.name).toBe('a');
    expect(res.modelType).toBe('enum-field');
    expect(res.list).toBeTruthy();
    expect((res.type as EnumType).type).toBe('enum');
    expect(res).toMatchSnapshot('default');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('default enum multiplicity == one', () => {
    const res = new SimpleField({
      name: 'A',
      type: {
        type: 'enum',
        name: 'ENUM1',
      },
    });
    expect(res.name).toBe('a');
    expect(res.modelType).toBe('enum-field');
    expect(res.list).toBeFalsy();
    expect((res.type as EnumType).type).toBe('enum');
    expect((res.type as EnumType).multiplicity).toBe('one');
    expect(res).toMatchSnapshot('default');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
});
