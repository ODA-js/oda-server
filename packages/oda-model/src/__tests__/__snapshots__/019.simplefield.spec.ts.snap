// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleField default enum multiplicity == one: default 1`] = `
SimpleField {
  Symbol(internal): Object {
    "args": Map {},
    "list": false,
    "name": "a",
    "type": Object {
      "multiplicity": "one",
      "name": "ENUM1",
      "type": "enum",
    },
  },
  Symbol(metadata): Object {
    "acl": Object {
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": true,
      "required": false,
    },
  },
}
`;

exports[`SimpleField default enum multiplicity == one: toObject 1`] = `
Object {
  "args": Object {},
  "defaultValue": undefined,
  "derived": false,
  "description": undefined,
  "entity": undefined,
  "identity": false,
  "indexed": false,
  "inheritedFrom": undefined,
  "list": false,
  "metadata": Object {
    "acl": Object {
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": true,
      "required": false,
    },
  },
  "name": "a",
  "order": undefined,
  "persistent": true,
  "required": false,
  "title": undefined,
  "type": Object {
    "multiplicity": "one",
    "name": "ENUM1",
    "type": "enum",
  },
}
`;

exports[`SimpleField default enum: default 1`] = `
SimpleField {
  Symbol(internal): Object {
    "args": Map {},
    "list": true,
    "name": "a",
    "type": Object {
      "multiplicity": "many",
      "name": "ENUM1",
      "type": "enum",
    },
  },
  Symbol(metadata): Object {
    "acl": Object {
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": true,
      "required": false,
    },
  },
}
`;

exports[`SimpleField default enum: toObject 1`] = `
Object {
  "args": Object {},
  "defaultValue": undefined,
  "derived": false,
  "description": undefined,
  "entity": undefined,
  "identity": false,
  "indexed": false,
  "inheritedFrom": undefined,
  "list": true,
  "metadata": Object {
    "acl": Object {
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": true,
      "required": false,
    },
  },
  "name": "a",
  "order": undefined,
  "persistent": true,
  "required": false,
  "title": undefined,
  "type": Object {
    "multiplicity": "many",
    "name": "ENUM1",
    "type": "enum",
  },
}
`;

exports[`SimpleField default: default 1`] = `
SimpleField {
  Symbol(internal): Object {
    "args": Map {},
    "list": false,
    "name": "a",
    "type": "string",
  },
  Symbol(metadata): Object {
    "acl": Object {
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": true,
      "required": false,
    },
  },
}
`;

exports[`SimpleField default: toObject 1`] = `
Object {
  "args": Object {},
  "defaultValue": undefined,
  "derived": false,
  "description": undefined,
  "entity": undefined,
  "identity": false,
  "indexed": false,
  "inheritedFrom": undefined,
  "list": false,
  "metadata": Object {
    "acl": Object {
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": true,
      "required": false,
    },
  },
  "name": "a",
  "order": undefined,
  "persistent": true,
  "required": false,
  "title": undefined,
  "type": "string",
}
`;
